src/app/GitUI/UserControls/FileStatusList.cs

```diff
@@ -289,40 +295,48 @@ private void SetupUnifiedDiffListSorting()
             _diffListSortSubscription?.Dispose();
 
             _diffListSortSubscription = DiffListSortService.Instance.CurrentAndFutureSorting()
-                .Do(sortingMethod =>
+                .Do(sortType =>
                 {
-                    switch (sortingMethod)
+                    _groupBy = sortType switch
                     {
-                        case DiffListSortType.FilePath:
-                            SortByFilePath();
-                            break;
-
-                        case DiffListSortType.FileExtension:
-                            SortByFileExtension();
-                            break;
+                        DiffListSortType.FilePath or DiffListSortType.FilePathFlat
+                            => null,
+                        DiffListSortType.FileExtension or DiffListSortType.FileExtensionFlat
+                            => new GroupBy(status => status.IsRangeDiff ? "~" : Path.GetExtension(status.Name), GetImageKey: _ => nameof(Images.File), GetLabel: k => k),
+                        DiffListSortType.FileStatus or DiffListSortType.FileStatusFlat
+                            => new GroupBy(GetStatusKey, GetImageKey: GetItemImageKey, GetLabel: _ => ""),
+                        _ => throw new NotSupportedException($"{sortType} is not a supported sorting method.")
+                    };
 
-                        case DiffListSortType.FileStatus:
-                            SortByFileStatus();
-                            break;
+                    _flatList = sortType.ToString().EndsWith("Flat");
 
-                        default:
-                            throw new NotSupportedException(sortingMethod.ToString() + " is not a supported sorting method.");
-                    }
+                    UpdateFileStatusListView(GitItemStatusesWithDescription, updateCausedByFilter: true);
                 })
-                .Catch<DiffListSortType, NotSupportedException>(e =>
+                .Catch<DiffListSortType, Exception>(ex =>
                 {
-                    // TODO log the error can we display it to the user somehow?
+                    Trace.WriteLine(ex);
                     return Observable.Empty<DiffListSortType>();
                 })
                 .Subscribe();
+
+            return;
+
+            static string GetStatusKey(GitItemStatus status)
+            {
+                char inverseDiffStatus = (char)((int)'Z' - (int)status.DiffStatus);
+                int comparableImageIndex = 10 + _imageListData.StateImageIndexMap[GetItemImageKey(status)];
+                DebugHelpers.Assert(inverseDiffStatus >= 0, $"offset of {nameof(inverseDiffStatus)} needs to be adapted");
+                DebugHelpers.Assert(comparableImageIndex < 100, $"offset of {nameof(comparableImageIndex)} needs to be adapted");
+                return $"{inverseDiffStatus}{comparableImageIndex}";
```

Comment:
Avoid the magic number 10
```suggestion
                return $"{inverseDiffStatus}{comparableImageIndex:D2}";
```